gossip/components/requester.c

```diff
@@ -17,136 +16,184 @@
                          tangle_t *const tangle) {
-  if (state == NULL) {
+  if (state == NULL || tangle == NULL) {
     return RC_REQUESTER_COMPONENT_NULL_STATE;
-  } else if (tangle == NULL) {
-    return RC_REQUESTER_COMPONENT_NULL_TANGLE;
   }
 
   logger_helper_init(REQUESTER_COMPONENT_LOGGER_ID, LOGGER_DEBUG, true);
   memset(state, 0, sizeof(requester_state_t));
+  state->milestones = NULL;
+  state->transactions = NULL;
   state->tangle = tangle;
+  rw_lock_handle_init(&state->lock);
 
-  log_debug(REQUESTER_COMPONENT_LOGGER_ID, "Initializing requester list\n");
-  if (CL_INIT(trit_array_p, state->list, trit_array_cmp) != CL_SUCCESS) {
-    log_critical(REQUESTER_COMPONENT_LOGGER_ID,
-                 "Initializing requester list failed\n");
-    return RC_REQUESTER_COMPONENT_FAILED_INIT_LIST;
-  }
   return RC_OK;
 }
 
-size_t requester_size(requester_state_t *const state) {
+retcode_t requester_destroy(requester_state_t *const state) {
   if (state == NULL) {
-    return 0;
+    return RC_REQUESTER_COMPONENT_NULL_STATE;
   }
 
-  return CL_SIZE(state->list);
+  hash243_set_free(&state->milestones);
+  hash243_set_free(&state->transactions);
+  state->tangle = NULL;
+  rw_lock_handle_destroy(&state->lock);
+  logger_helper_destroy(REQUESTER_COMPONENT_LOGGER_ID);
+
+  return RC_OK;
 }
 
-retcode_t requester_clear_request(requester_state_t *const state,
-                                  trit_array_p const hash) {
-  if (state == NULL) {
+retcode_t requester_get_requested_transactions(
+    requester_state_t *const state, hash243_set_t *const transactions) {
+  retcode_t ret = RC_OK;
+
+  if (state == NULL || transactions == NULL) {
     return RC_REQUESTER_COMPONENT_NULL_STATE;
-  } else if (hash == NULL) {
-    return RC_REQUESTER_COMPONENT_NULL_HASH;
   }
 
-  if (CL_REMOVE(state->list, hash) != CL_SUCCESS) {
-    log_warning(REQUESTER_COMPONENT_LOGGER_ID,
-                "Clearing request from list failed\n");
-    return RC_REQUESTER_COMPONENT_FAILED_REMOVE_LIST;
+  rw_lock_handle_rdlock(&state->lock);
+
+  if ((ret = hash243_set_append(&state->transactions, transactions)) != RC_OK) {
+    goto done;
   }
-  return RC_OK;
+  if ((ret = hash243_set_append(&state->milestones, transactions)) != RC_OK) {
+    goto done;
+  }
+
+done:
+  rw_lock_handle_unlock(&state->lock);
+  return ret;
+}
+
+size_t requester_size(requester_state_t *const state) {
+  size_t size = 0;
+
+  if (state == NULL) {
+    return RC_REQUESTER_COMPONENT_NULL_STATE;
+  }
+
+  rw_lock_handle_rdlock(&state->lock);
+  size = hash243_set_size(&state->transactions) +
+         hash243_set_size(&state->milestones);
+  rw_lock_handle_unlock(&state->lock);
+
+  return size;
 }
 
 bool requester_is_full(requester_state_t *const state) {
+  size_t size = 0;
+
   if (state == NULL) {
     return RC_REQUESTER_COMPONENT_NULL_STATE;
   }
 
-  return CL_SIZE(state->list) >= MAX_TX_REQ_NBR;
+  rw_lock_handle_rdlock(&state->lock);
```

Comment:
`Thread safety`
`You can use uthash in a threaded program. But you must do the locking. Use a read-write lock to protect against concurrent writes. It is ok to have concurrent readers (since uthash 1.5).`

Reply:
we are on 2.0.2

Reply:
After discussion/comprehension of a pretty elusive documentation, we'll let the `rw_lock_handle_rdlock `.